File: ios/Classes/HsiBuilder.swift
import Foundation

/// Builds HSI 1.0 JSON maps from HR samples. Mock implementation matching the Dart MockSessionEngine.
enum HsiBuilder {

    /// A single HR sample with timestamp and BPM value.
    typealias HrSample = (timestampMs: Int64, bpm: Double)

    /// Build an HSI 1.0 dict from HR samples.
    ///
    /// - Parameters:
    ///   - samples: Array of (timestampMs, bpm) tuples.
    ///   - config: The active session configuration.
    ///   - seq: Sequence number of this HSI frame.
    /// - Returns: A dictionary matching the Dart `hsi_json` wire format.
    static func build(samples: [HrSample], config: SessionConfig, seq: Int) -> [String: Any] {
        guard !samples.isEmpty else {
            return emptyHsi(config: config, seq: seq)
        }

        let startMs = samples.first!.timestampMs
        let endMs = samples.last!.timestampMs

        // Compute mean BPM
        let bpmValues = samples.map { $0.bpm }
        let meanBpm = bpmValues.reduce(0, +) / Double(bpmValues.count)

        // Convert BPM to RR intervals (ms)
        let rrIntervals = bpmValues.map { 60000.0 / $0 }
        let meanRR = rrIntervals.reduce(0, +) / Double(rrIntervals.count)

        // SDNN: population standard deviation of RR intervals
        let variance = rrIntervals.reduce(0.0) { $0 + ($1 - meanRR) * ($1 - meanRR) }
            / Double(rrIntervals.count)
        let sdnn = sqrt(variance)

        // RMSSD approximation (matches Dart mock: sdnn * 0.8)
        let rmssd = sdnn * 0.8

        let window: [String: Any] = [
            "start_ms": startMs,
            "end_ms": endMs,
            "sample_count": samples.count
        ]
        let physio: [String: Any] = [
            "hr_mean_bpm": round(meanBpm * 10) / 10,
            "hr_sdnn_ms": round(sdnn * 100) / 100,
            "rmssd_ms": round(rmssd * 100) / 100
        ]
        let meta: [String: Any] = [
            "mock": true,
            "session_id": config.sessionId,
            "mode": config.mode.rawValue,
            "seq": seq
        ]

        return [
            "hsi_version": "1.0",
            "producer": "synheart-session-ios/0.1.0",
            "windows": [window],
            "axes": ["physiological": physio],
            "privacy": ["raw_samples_included": false],
            "meta": meta
        ]
    }

    private static func emptyHsi(config: SessionConfig, seq: Int) -> [String: Any] {
        let physio: [String: Any] = [
            "hr_mean_bpm": 0.0,
            "hr_sdnn_ms": 0.0,
            "rmssd_ms": 0.0
        ]
        let meta: [String: Any] = [
            "mock": true,
            "session_id": config.sessionId,
            "mode": config.mode.rawValue,
            "seq": seq
        ]

        return [
            "hsi_version": "1.0",
            "producer": "synheart-session-ios/0.1.0",
            "windows": [] as [[String: Any]],
            "axes": ["physiological": physio],
            "privacy": ["raw_samples_included": false],
            "meta": meta
        ]
    }
}

File: android/src/main/kotlin/ai/synheart/session/HsiBuilder.kt
package ai.synheart.session

import kotlin.math.sqrt
import kotlin.math.round

/** Builds HSI 1.0 JSON maps from HR samples. Mock implementation matching the Dart MockSessionEngine. */
object HsiBuilder {

    /**
     * Build an HSI 1.0 map from HR samples.
     *
     * @param samples List of (timestampMs, bpm) pairs.
     * @param config The active session configuration.
     * @param seq Sequence number of this HSI frame.
     * @return A map matching the Dart `hsi_json` wire format.
     */
    fun build(samples: List<Pair<Long, Double>>, config: SessionConfig, seq: Int): Map<String, Any> {
        if (samples.isEmpty()) return emptyHsi(config, seq)

        val startMs = samples.first().first
        val endMs = samples.last().first

        // Compute mean BPM
        val bpmValues = samples.map { it.second }
        val meanBpm = bpmValues.sum() / bpmValues.size

        // Convert BPM to RR intervals (ms)
        val rrIntervals = bpmValues.map { 60000.0 / it }
        val meanRR = rrIntervals.sum() / rrIntervals.size

        // SDNN: population standard deviation of RR intervals
        val variance = rrIntervals.sumOf { (it - meanRR) * (it - meanRR) } / rrIntervals.size
        val sdnn = sqrt(variance)

        // RMSSD approximation (matches Dart mock: sdnn * 0.8)
        val rmssd = sdnn * 0.8

        return mapOf(
            "hsi_version" to "1.0",
            "producer" to "synheart-session-android/0.1.0",
            "windows" to listOf(
                mapOf(
                    "start_ms" to startMs,
                    "end_ms" to endMs,
                    "sample_count" to samples.size
                )
            ),
            "axes" to mapOf(
                "physiological" to mapOf(
                    "hr_mean_bpm" to round(meanBpm * 10) / 10,
                    "hr_sdnn_ms" to round(sdnn * 100) / 100,
                    "rmssd_ms" to round(rmssd * 100) / 100
                )
            ),
            "privacy" to mapOf(
                "raw_samples_included" to false
            ),
            "meta" to mapOf(
                "mock" to true,
                "session_id" to config.sessionId,
                "mode" to config.mode.value,
                "seq" to seq
            )
        )
    }

    private fun emptyHsi(config: SessionConfig, seq: Int): Map<String, Any> = mapOf(
        "hsi_version" to "1.0",
        "producer" to "synheart-session-android/0.1.0",
        "windows" to emptyList<Map<String, Any>>(),
        "axes" to mapOf(
            "physiological" to mapOf(
                "hr_mean_bpm" to 0.0,
                "hr_sdnn_ms" to 0.0,
                "rmssd_ms" to 0.0
            )
        ),
        "privacy" to mapOf(
            "raw_samples_included" to false
        ),
        "meta" to mapOf(
            "mock" to true,
            "session_id" to config.sessionId,
            "mode" to config.mode.value,
            "seq" to seq
        )
    )
}
